var/spack/repos/builtin/packages/py-scipy/package.py

```diff
@@ -57,17 +58,20 @@ class PyScipy(PythonPackage):
     depends_on("python@:3.9", when="@1.5.4:1.7.1", type=("build", "link", "run"))
     depends_on("python@:3.8", when="@1.3.2:1.5.3", type=("build", "link", "run"))
 
-    depends_on("py-meson-python@0.12.1:", when="@1.11:", type="build")
-    depends_on("py-meson-python@0.11:", when="@1.10:", type="build")
-    depends_on("py-meson-python@0.9:", when="@1.9.2:", type="build")
-    depends_on("py-meson-python@0.8.1:", when="@1.9.1:", type="build")
-    depends_on("py-meson-python@0.7:", when="@1.9:", type="build")
+    depends_on("py-meson-python@0.15", when="@1.12:", type="build")
```

Comment:
@rgommers in the past I know we talked about leaving these upper bounds open-ended. However, Spack now has support for separate concretization of build-only dependencies similar to pip. So I don't think it's necessary, we can pin to exact versions now if we want.

Reply:
What exactly is the benefit of an upper bound and why is that better at build time than at runtime?

Note that pip does NOT pin build time dependencies by default and the only way to do it is to release new tagged upstream point releases every time you want to change a pin. So I would argue there is precisely zero prior art here, even if spack contains the technological capability to update pins without new point releases of e.g. a package build ID.

Reply:
That's not a question for me, it's a question for the scipy devs. Most build dependencies have historically been pinned to the latest release for "stability" reasons.

Reply:
Well, this goes back to the pinning method being part of the project source code and consequently part of the release tag as well.

- It's not possible to optimistically allow newer versions on the assumption that it most likely works, and adapt to the odd case of it not working, by adding a pin. You get one chance to publish your source code tarball to PyPI, and that's at the initial time of release.
- for cibuildwheel, you may want to impose specific restrictions, most famously the use of oldest-supported-numpy / a lower bound on numpy, which aren't even relevant *at all* to people building numpy for personal use, only relevant to cibuildwheel. But the PyPA ecosystem doesn't really have an official pinning infrastructure, and there's no official way to pin dependencies "for cibuildwheel", so lower/upper bounds in build dependencies in pyproject.toml is used as a "pragmatic fallback" to get the official cibuildwheel artifact builds to produce predictable results that support a wide range of numpy ABIs.

Spack doesn't operate under any of these restrictions.

Reply:
> However, Spack now has support for separate concretization of build-only dependencies similar to pip.

Nice to see this improvement!

> Well, this goes back to the pinning method being part of the project source code and consequently part of the release tag as well.
> [...] Spack doesn't operate under any of these restrictions.

Yes, you're correct about the rationale upstream @eli-schwartz. I really wish PyPI had a better model here, but it doesn't. So for Spack the choice here is:
1. do I simply copy the upstream upper bounds, and loosen them in case there is an issue with an _older_ version of a build dependency
2. do I ignore the upstream upper bounds, and add them whenever there is an issue with a _newer_ version of a build dependency

Given that build deps are concretized separately, the choice only matters for building `py-scipy`. And I don't think the answer matters all that much, it's just whatever is expected to be less effort. And that's hard to tell ahead of time; it's probably similar for both cases. So I'm fine with either choice.

Reply:
Heh. To be fair by "here" I meant "derived from the pip comparison".

Reply:
> To be fair by "here" I meant "derived from the pip comparison".

Okay, then fair enough.

> Also I disagree with cargo's reasoning

Yeah, I think it's not that hard to produce a counter-example. That said, they're correct for >=99% of the time. `glibc` is about as stable as it gets - but still, it's got lots of bugs, and NumPy contains a bunch of workarounds for them. So definitely not every version works; one needs at least a lower bound for nontrivial usage. Even bug fixes in newer versions may be bc-breaking.

Reply:
> Spack doesn't operate under any of these restrictions.

This is a very valid point.

> So for Spack the choice here is:

And a great summary of the pros and cons of each.

In that sense, I wonder if the lower bound is even a real lower bound. Should we leave off all version requirements until we know for certain that they are real?

I'm fine with both options honestly. @rgommers has been great about responding to Spack issues for numpy/scipy, so I'm not particularly worried about not keeping up-to-speed with issue reports. I'll let y'all decide which you prefer.

Reply:
> In that sense, I wonder if the lower bound is even a real lower bound. Should we leave off all version requirements until we know for certain that they are real?

Almost all lower bounds are actual bounds. Usually what happens is that a lower bound is bumped when a real-world bug shows that the older version doesn't work. The only thing you cannot tell from the metadata is whether that bug is a general incompatibility (e.g., the package started using a new feature in the dependency that isn't supported in versions below the lower bound) or it's a platform-specific bug.

> I'll let y'all decide which you prefer.

Now that Spack separately concretizes build deps, really either one works. So I don't have a super strong preference either way. However, I do know that the majority of distro maintainers prefer no upper bounds (as does @eli-schwartz in this discussion), so I suggest we go with that. To be explicit, I propose the following policy for taking over version constraints from metadata in `pyproject.toml` files:

For `build-system.requires`:

1. Do take over lower bounds
2. Do not take over upper bounds, unless there is a specific comment in pyproject.toml that this upper bound is necessary for reason X, and that reason applies in Spack

For `project.dependencies`: take over all constraints exactly.

Reply:
Works for me. Let me update both PRs with this new policy.